### DISCUSSION TITLE
Trouble setting `assets` with custom MultiBaseReader

### DISCUSSION THREAD

**Discussion by mccarthyryanc** [cogeotiff/rio-tiler#720] [Q&A]
I have a custom MultiBaseReader that looks similar to this (much like STACReader), simplified a bit just to make this clear:
```python
from typing import Type, Dict, Any

import attr
import rasterio
from rio_tiler.io.base import BaseReader, MultiBaseReader
from rio_tiler.io.rasterio import Reader
from rio_tiler.types import AssetInfo


@attr.s
class CustomReader(MultiBaseReader):

    def __attrs_post_init__(self):
        self.assets = ["/path/to/raster.tif"]
        self.bounds = [114.101509, 22.437414, 114.156379, 22.482232]
        self.crs = rasterio.crs.CRS.from_epsg(4326)
    
    def _get_asset_info(self, asset: str) -> AssetInfo:
        return AssetInfo(url=self.assets[0])
```

I am instantiating and calling `tile` (with x,y,z values that I know overlap the raster):
```python
r = CustomReader(input="bogus")
r.tile_x=26772, tile_y=14282, tile_z=15)
```

But I get this error:
```
MissingAssets: assets must be passed either via `expression` or `assets` options.
```

Is there a way to pass assets via `self.assets` and have `tile` use them? Or do I need to update the `tile` method to assign from assets similar to `info` and `statistics` methods?:
```python
assets = assets or self.assets
```

**Comment by vincentsarago** [comment]
@mccarthyryanc I'm not sure why you're using `MultiBaseReader` for this or maybe you're example is just too much reduced 

a `MultiBaseReader` is designed to with an input dataset in format of `{"asset1": "url1", "asset2": url2}` and the `assets` list should be a list of asset keys `["asset1", "asset2"]` 

By default the `tile()` method will expect to have an `assets` option passed which will then be used to get the URL. So if I understand well, you want the `tile()` method to default to all assets? This could be quite dangerous for the majority of users (IMO). I will work well when you have a couple assets but if you have a STAC Item with 10 assets (with them having multiple bands) this could results in weird things. 

Maybe we can introduce a `default_assets` attribute

**Reply by mccarthyryanc** [reply]
Yeah, sorry a little oversimplified :sweat_smile:  I have an internal STAC-like API that I'm trying to work with, basically the same as the `STACReader` with a custom `fetch` function and a fixed asset list that I'd need to access at a time. That's why I was using `MultiBaseReader`.

A `default_assets` attribute would be helpful if it could read from `self.assets`. Where my `CustomReader` would define the `default_assets` and then `tile` (and the other similar methods part/point/etc.) would maybe do something like:
```python
   def tile(
        self,
        tile_x: int,
        tile_y: int,
        tile_z: int,
        assets: Union[Sequence[str], str] = None,
        expression: Optional[str] = None,
        asset_indexes: Optional[Dict[str, Indexes]] = None,  # Indexes for each asset
        asset_as_band: bool = False,
        **kwargs: Any,
    ) -> ImageData:
...
        if expression:
            assets = self.parse_expression(expression, asset_as_band=asset_as_band)

        assets = assets or self.default_assets

        if not assets:
            raise MissingAssets(
                "assets must be passed via `expression`, `assets` options, or the class-level `default_assets`."
            )
...
```

If that sounds okay, I'd be happy putting together a PR.

**Reply by vincentsarago** [reply]
are you using `titiler-pgstac`?

**Reply by mccarthyryanc** [reply]
I'm not, just plain `titiler`. But I did look at that repo as an option. I'm not querying against postgres, so it seemed like that project did more than what I needed.

**Comment by mccarthyryanc** [comment]
I ended up creating a [PR](https://github.com/cogeotiff/rio-tiler/pull/722) with a solution that is working really well for me. I added the attribute to `STACReader` so it was simpler to test and document. But, I totally understand if this isn't great to put in `STACReader` and happy to adapt.
